tests/process.md

```diff
@@ -0,0 +1,159 @@
+# Setting up the environment
+
+```ocaml
+# #require "eio_main";;
+```
+
+Creating some useful helper functions
+
+```ocaml
+open Eio.Std
+
+module Flow = Eio.Flow
+module Process = Eio.Process
+
+let ( / ) = Eio.Path.( / )
+
+let run fn =
+  Eio_main.run @@ fun env ->
+  fn env#process_mgr env
+
+let check_signal expected = function
+  | `Signaled x when x = expected -> Ok ()
+  | x -> Error x;;
+```
+
+Running a program as a subprocess:
+
+```ocaml
+# run @@ fun mgr _env ->
+  Switch.run @@ fun sw ->
+  let t = Process.spawn ~sw mgr [ "echo"; "hello world" ] in
+  Process.await t;;
+hello world
+- : Process.exit_status = `Exited 0
+```
+
+Stopping a subprocess works and checking the status waits and reports correctly:
+
+```ocaml
+# run @@ fun mgr _env ->
+  Switch.run @@ fun sw ->
+  let t = Process.spawn ~sw mgr [ "sleep"; "10" ] in
+  Process.signal t Sys.sigkill;
+  Process.await t |> check_signal Sys.sigkill;;
+- : (unit, Process.exit_status) result = Ok ()
+```
+
+A switch will stop a process when it is released:
+
+```ocaml
+# run @@ fun mgr env ->
+  let proc = Switch.run (fun sw -> Process.spawn ~sw mgr [ "sleep"; "10" ]) in
+  Process.await proc |> check_signal Sys.sigkill;;
+- : (unit, Process.exit_status) result = Ok ()
+```
+
+Passing in flows allows you to redirect the child process' stdout:
+
+```ocaml
+# run @@ fun mgr env ->
+  let fs = Eio.Stdenv.fs env in
+  let filename = "process-test.txt" in
+  Eio.Path.(with_open_out ~create:(`Exclusive 0o600) (fs / filename)) @@ fun stdout ->
+  Process.run mgr ~stdout [ "echo"; "Hello" ];
+  Eio.Path.(load (fs / filename));;
+- : string = "Hello\n"
+```
+
+Piping data to and from the child:
+
+```ocaml
+# run @@ fun mgr env ->
+  let stdin = Eio.Flow.string_source "one\ntwo\nthree\n" in
+  Process.parse_out mgr Eio.Buf_read.line ~stdin ["wc"; "-l"] |> String.trim;;
+- : string = "3"
+```
+
+Spawning subprocesses in new domains works normally:
+
+```ocaml
+# run @@ fun mgr env ->
+  Eio.Domain_manager.run env#domain_mgr @@ fun () ->
+  Process.run mgr [ "echo"; "Hello from another domain" ];;
+Hello from another domain
+- : unit = ()
+```
+
+Calling `await_exit` multiple times on the same spawn just returns the status:
+
+```ocaml
+# run @@ fun mgr env ->
+  Switch.run @@ fun sw ->
+  let t = Process.spawn ~sw mgr [ "echo"; "hello world" ] in
+  (Process.await t, Process.await t, Process.await t);;
+hello world
+- : Process.exit_status * Process.exit_status * Process.exit_status =
+(`Exited 0, `Exited 0, `Exited 0)
+```
+
+Using a sink that is not backed by a file descriptor:
+
+```ocaml
+# run @@ fun mgr env ->
+  let buf = Buffer.create 16 in
+  Eio.Process.run mgr ~stdout:(Flow.buffer_sink buf) [ "echo"; "Hello, world" ];
+  Buffer.contents buf
+- : string = "Hello, world\n"
+```
+
+Changing directory (unconfined):
+
+```ocaml
+# run @@ fun mgr env ->
+  let root = env#fs / "/" in
+  Process.run mgr ~cwd:root [ "env"; "pwd" ];;
+/
+- : unit = ()
+```
+
+Changing directory (confined):
```

Comment:
I'm pretty confident the code is correct, but do you think we need an example that tries to access a directory it shouldn't ?

Reply:
I added a test to check the error is sensible. It probably doesn't matter too much; either the OS is restricting it (in which case checking isn't our problem), or the OS will let the child access things outside of it anyway.

Reply:
Indeed, thanks!